### ISSUE TITLE
Restrict position of tooltips to within a certain viewport

### ISSUE THREAD

**Issue by SimonVanneste** [plouc/nivo#2403]
There seems to be no way at the moment to manipulate the calculated position so the tooltip doesn't get cut off at the edges of the "viewport" (tooltip container, custom viewport or `window`):
![Screenshot 2023-07-28 at 14 31 19](https://github.com/plouc/nivo/assets/38662306/f9aab876-2c29-403e-ba6e-8271a85d1a04)

Right now, tooltips are positioned using `transform: translate(...)` inside `TooltipWrapper`: 
https://github.com/plouc/nivo/blob/2671ca4df614190f597bb63030bb794d3106c36e/packages/tooltip/src/TooltipWrapper.tsx#L39-L69 

Some implementation options:

- User-land: add a `transformTooltipPosition(x: number, y: number, tooltipWidth: number, tooltipHeight: number)` prop and allow users of this library to implement their own logic.
- Inside library: add a `TooltipViewport` context that can optionally be rendered at the desired location by library users and implement the positioning logic inside the library (keep the current behaviour if no context is available).
  ```tsx
  <TooltipViewport.Provider value={viewportRef}>
    <div ref={viewportRef}>
      ...
         <Bar ... />
      ...
    </div>
  </TooltipViewport.Provider value={viewportRef}>
  ```

I'd be willing to help with the implementation if we land on an option.

**Comment by ollwenjones** [comment]
Would love to see this. Seems like keeping it inside the current viewport, so it adjusts its position instead of getting cut off, would solve this most of the time. Otherwise, it seems like the actual `overflow:hidden` point being the edge of the window, not the bounding-box of the chart itself, would solve a lot of real-life scenarios. A combo of the two would be ideal, imo.

**Comment by jamesarosen** [comment]
Is `TooltipViewport.Provider` really required? Couldn't the library use [createPortal](https://react.dev/reference/react-dom/createPortal) to render the tooltip content as a child of `<body>` to avoid overflow issues?

**Comment by nm2501** [comment]
I'm also keen to see a solution here. Great suggestions @SimonVanneste . 
My preference would be the `transformTooltipPosition` option but perhaps with more parameters so that the callback doesn't need to determine the overall chart size itself. I think the library already knows the chart size so it would be easy to pass that to the transform callback.

**Comment by ollwenjones** [comment]
@jamesarosen yes, a portal would be better for when the edge of the chart is the problem, but the tooltip still needs to measure against the viewport to adjust its position, so it doesn't go off the edge of the browser window when an element is near the edge of the window.

**Comment by fsobanski** [comment]
bump

**Comment by KissBalazs** [comment]
bump

**Comment by vincent-azumuta** [comment]
bump

**Comment by sebastianoscarlopez** [comment]
same issue here, is there any work around for this?

<img width="324" alt="image" src="https://github.com/plouc/nivo/assets/10349981/01327a8d-f303-489e-a82a-cd07f1f5a003">

**Comment by musetraine** [comment]
hello, have the same situation, how I can fix it?
<img width="1633" alt="image" src="https://github.com/plouc/nivo/assets/18230499/4d683b40-3b78-4b2a-af38-6c353cb1448d">

**Comment by sebastianoscarlopez** [comment]
I was able to fix it using Popper's VirtualElement. It is not a generic solution, but it worked for me.

**Comment by musetraine** [comment]
> I was able to fix it using Popper's VirtualElement. It is not a generic solution, but it worked for me.

Could you share the code please?

**Comment by stolium** [comment]
Hey folks! Im totally new in floating ui but this one working for me:

**chart**
```
<ResponsiveBar
tooltip={Tooltip}
```

**Tooltip.tsx**
```
import { flip, shift, useFloating } from '@floating-ui/react';

export const Tooltip = () => {
const { refs, floatingStyles } = useFloating({
    placement: 'right-start',
    middleware: [flip(), shift()],
  });

  return (
    <TooltipWrapper position={[0, 0]} anchor="center">
      <div ref={refs.setReference}/>
      <FloatingTooltip
        ref={refs.setFloating}
        floatingStyles={floatingStyles}
      >
        content
     </FloatingTooltip>
   </TooltipWrapper>);
   }
```

**FloatingTooltip.tsx**
```
type CustomTooltipProps = {
  floatingStyles: any;
};

export const FloatingTooltip = forwardRef<
  HTMLDivElement,
  PropsWithChildren<CustomTooltipProps>
>(function CustomTooltip({ children, floatingStyles }, ref) {
  return (
    <FloatingPortal>
      <div
        ref={ref}
        className={cn([styles.customTooltip, 'fadeIn'])}
        style={floatingStyles}
      >
        {children}
      </div>
    </FloatingPortal>
  );
});
```

So idea behind this it to place anchor element to original position inside `<TooltipWrapper>` with `<div ref={refs.setReference}></div>` and then Floating ui take care of everything by applying  middleware: [flip(), shift()].

**Comment by IoghiStudio** [comment]
> Hey folks! Im totally new in floating ui but this one works for me:
> 
> **chart**
> 
> ```
> <ResponsiveBar
> tooltip={Tooltip}
> ```
> 
> **Tooltip.tsx**
> 
> ```
> import { flip, shift, useFloating } from '@floating-ui/react';
> 
> export const Tooltip = () => {
> const { refs, floatingStyles } = useFloating({
>     placement: 'right-start',
>     middleware: [flip(), shift()],
>   });
> 
>   return (
>     <TooltipWrapper position={[0, 0]} anchor="center">
>       <div ref={refs.setReference}/>
>       <FloatingTooltip
>         ref={refs.setFloating}
>         floatingStyles={floatingStyles}
>       >
>         content
>      </FloatingTooltip>
>    </TooltipWrapper>);
>    }
> ```
> 
> **FloatingTooltip.tsx**
> 
> ```
> type CustomTooltipProps = {
>   floatingStyles: any;
> };
> 
> export const FloatingTooltip = forwardRef<
>   HTMLDivElement,
>   PropsWithChildren<CustomTooltipProps>
> >(function CustomTooltip({ children, floatingStyles }, ref) {
>   return (
>     <FloatingPortal>
>       <div
>         ref={ref}
>         style={floatingStyles}
>       >
>         {children}
>       </div>
>     </FloatingPortal>
>   );
> });
> ```
> 
> So idea behind this it to place anchor element to original position inside `<TooltipWrapper>` with `<div ref={refs.setReference}/>` and then Floating ui take care of everything by applying middleware: [flip(), shift()]. [@vengalraoApperture](https://github.com/vengalraoApperture) [@musetraine](https://github.com/musetraine) [@fsobanski](https://github.com/fsobanski)

@stolium Where is the TooltipWrapper from ? If is custom can you share it ?

**Comment by thehighestprimenumber** [comment]
bump

**Comment by zhikin2207** [comment]
bump

**Comment by zhikin2207** [comment]
@musetraine, your case should be easy to fix, just pass the tooltip z-index. 

```
          // @ts-ignore
          theme={{ tooltip: { zIndex: 10000 } }}
```

**Comment by Ziker22** [comment]
Hey folks, i was able to fix the problem via vanilla react.

Im using styled-components, but swapping it for a different styling solution should be trivial.

`CHART_BOX_SELECTOR` is a basic HTML element - in my case a tile that holds the chart  - swap for the elemet in your app whose borders should be respected by the tooltip
```
import React, { FC, PropsWithChildren, useEffect, useRef, useState } from 'react';
import styled from 'styled-components';

// Generic hook that tracks mouse position
import useMousePosition from '../../../hooks/useMousePosition';

const CHART_BOX_SELECTOR = '[data-testid="dashboard-widget"]';

export const TranslatedTooltip = styled.div<{ movePx: number }>`
  // additional styles for tooltip
  transform: ${({ movePx }) => `translateX(${movePx}px)`};
  transition: transform;
  & ::after {
    // additional styles for tooltip arrow
    transform: ${({ movePx }) => `translateX(${-movePx - (movePx > 0 ? -7 : 7)}px)`};
    content: ' ';
    position: absolute;
    top: 100%;
    left: 50%;
    margin-left: -5px;
  }
`;

export const ChartTooltipPositioner: FC<PropsWithChildren> = ({ children }) => {
  const ref = useRef<HTMLDivElement>(null);
  const mousePosition = useMousePosition();
  const [tooltipXTranslation, setTooltipXTranslation] = useState(0);
  useEffect(() => {
    const adjustTooltipPosition = () => {
      const tooltip = ref.current;
      if (!tooltip) return;

      const parent = tooltip.closest(CHART_BOX_SELECTOR);
      if (!parent) return;

      const tooltipBox = tooltip.getBoundingClientRect();
      const parentBox = parent.getBoundingClientRect();

      const isStartOutOfBounds = tooltipBox.left < parentBox.left;
      const isEndOutOfBounds = tooltipBox.right > parentBox.right;

      if (isStartOutOfBounds || isEndOutOfBounds) {
        const modifier = isEndOutOfBounds ? -1 : 1;
        setTooltipXTranslation(modifier * (tooltipBox.width / 2));
      } else {
        setTooltipXTranslation(0);
      }
    };

    adjustTooltipPosition();
  }, [mousePosition]);

  return (
    <div ref={ref}>
      <TranslatedTooltip movePx={tooltipXTranslation}>{children}</TranslatedTooltip>
    </div>
  );
};

// Actual Chart

const MyLineChart = () => {

return <ResponsiveLine
...
tooltip={() =><ChartTooltipPositioner>Your content here</ChartTooltipPositioner>}
...
/>
}

```

**Comment by hsl0505** [comment]
bump
